code/scripts/DirectoryController.hs

```diff
@@ -0,0 +1,112 @@
+
+-- iterates through drasil- package; outputs subdirectories and haskell files
+iterator :: Folder -> IO ([Folder],[File])
+iterator folder = do
+  setCurrentDirectory (folderPath folder)
+  rawContents <- listDirectory (folderName folder)
+
+  let bakedContents = (folders,files)
+      folders = map (createFolder workingDirectory currentDrasilPack) folderNames
+      files = map (createFile workingDirectory currentDrasilPack) fileNames
+
+      folderNames = sort $ rawContents \\ filter (isInfixOf ".") rawContents
+      fileNames = sort $ filter (isSuffixOf ".hs") rawContents
+
+      workingDirectory = getFolderPath folder
+      currentDrasilPack = folderDrasilPack folder
+  return bakedContents
+
+-- recursively searches for all folders + files in a directory using iterator
+finder :: Folder -> IO [File]
+finder folder = do
+  rawData <- iterator folder
+
+  let rawFolders = fst rawData
+  folders <- verifyDirectories rawFolders
+  rawFiles <- mapM finder folders
+
+  let bakedFiles
+        | null rawFiles = []
+        | otherwise = concat rawFiles
+
+  let files 
+        | null folders = snd rawData
+        | otherwise = bakedFiles ++ snd rawData
+  return files
+
+-- gets all drasil- packages + filepaths in a list of folder data types
+getDirectories :: FilePath -> FilterPrefix -> IO [Folder]
+getDirectories directoryPath filterPrefix = do
+  -- all raw directory contents
+  all <- listDirectory directoryPath
+  -- raw drasil- package directories + package names
+  let rawPackages = sort $ filter (isPrefixOf filterPrefix) all
+      packageNames = map (\\"drasil-") rawPackages
+  -- convert list of directories into folder data types
+      directories = zipWith (createFolder directoryPath) packageNames rawPackages
+  return directories
+
+-- verifies that each folder/directory exists
+verifyDirectories :: [Folder] -> IO [Folder]
+verifyDirectories rawFolders = do
+  let rawDirectories = map getFolderPath rawFolders
+  boolFolders <- mapM doesDirectoryExist rawDirectories
+  let verifiedDirectories = snd $ partition nullFolder (zipWith fBool boolFolders rawFolders)
+  return verifiedDirectories
+
+-- combines lists of Booleans and Folders (if True, Folder exists)
+fBool :: Bool -> Folder -> Folder
+fBool b f = if b then f else createFolder "" "" ""
+
+-- checks if a folder is null (empty)
+nullFolder :: Folder -> Bool
+nullFolder folder = empty where
+  empty = all null [folderName folder,folderPath folder,folderDrasilPack folder]
+
+-- combines folder name with filepath (for testing if directory exists)
+joins :: [FilePath] -> FilePath
+joins (a:b:_) = b ++ "/" ++ a
+
+-- creates new folder path with folder name + path (to extract folder contents)
+getFolderPath :: Folder -> FilePath
+getFolderPath folder = folderPath folder ++ "/" ++ folderName folder
+
+-- strips pathSuffix from filepath, returning stripped filepath 
+-- strip occurs iff pathSuffix is suffix of filepath
+stripPath :: FilePath -> FilePath -> IO FilePath
```

Comment:
Is this function not pure?

Reply:
This function could be rewritten without the IO () part; My bad.

Reply:
No worries. These things can be easy to miss during refactors and code evolution. 🙂

Reply:
I refractored the function `stripPath` and removed the IO () part [here](https://github.com/JacquesCarette/Drasil/pull/2208/commits/4065330bf5885d36437fb97d322a5c23d339df61#) (old lines 106-112; new lines 101-106).